### PR TITLE
Allow passing optional assumed shape or pointers to optional explicit shape

### PR DESCRIPTION
**Pull request by jeanPerier**
Passing assumed shapes and pointers to explicit shape requires
copy-in/copy-out. This was already implemented, but did not account for
the fact the argument may be optional, in which case:

- If the argument is itself an optional, care must be taken to only
  read its descriptors if it is present. The copy-in/copy-out must
  be conditional on its presence.

- If the argument is a pointer, its descriptor can be read, but care
  must be taken not to allocate and pass a zero sized temp if it is
  un-allocated because it is fine to pass a disassociated pointer to
  a non-pointer optional argument, and the passed argument must be
  absent in this case. This point also applies to allocatables,
  although these are not problematic since they are contiguous and
  no copy-in/copy-out is done.

While writing the tests, it appeared allocatables were getting copied-in/copied-out. The root cause was found in semantics and a fix for it is part of this patch in its own commit that will be up-streamed directly to LLVM.

### CODE REVIEW

**Comment by jeanPerier** [comment]
IsSimplyContiguous fix merged in llvm here: https://reviews.llvm.org/D114599

**Review by psteinfeld — APPROVED** [review]
All builds, tests, and looks good to me.

I'm not that familiar with this code, though, so you might want to get an approval from a more knowledgeable person before merging.

File: flang/lib/Evaluate/check-expression.cpp
Comment (schweitzpgi) [inline flang/lib/Evaluate/check-expression.cpp:669]: Seems like a bug.

File: flang/lib/Lower/ConvertExpr.cpp
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertExpr.cpp:1842]: Dependent derived type is a TODO here?

File: flang/lib/Lower/ConvertExpr.cpp
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertExpr.cpp:2133]: Should this class be nested? Since it pertains to call handling which has code in another file thought I should ask.

File: flang/lib/Lower/ConvertExpr.cpp
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertExpr.cpp:2137]: What about arrays of other intrinsic types or scalar characters? Arrays should really be ArrayBoxValue.

File: flang/lib/Lower/ConvertExpr.cpp
Comment (schweitzpgi) [inline flang/lib/Lower/ConvertExpr.cpp:2205]: We probably need to sort out making copies of arrays and have a generic routine in fir::factory as we already make copies elsewhere.

**Review by schweitzpgi — APPROVED** [review]

**Review by jeanPerier — COMMENTED** [review]
Thanks for the review

**Pull request merged**
